src/Interpreters/ConcurrentHashJoin.cpp

```diff
@@ -301,6 +303,30 @@ bool ConcurrentHashJoin::addBlockToJoin(const Block & right_block_, bool check_l
     return true;
 }
 
+class ConcatStreams final : public IBlocksStream
+{
+public:
+    explicit ConcatStreams(std::vector<IBlocksStreamPtr> children_)
+        : children(std::move(children_)) {}
+
+    Block nextImpl() override
+    {
+        while (!children.empty())
+        {
+            auto & child = children.front();
+            if (!child)
+                continue;
+            Block b = child->next();
+            if (!b.empty()) return b;
+              children.erase(children.begin());
+        }
+        return {};
+    }
+
+private:
+    std::vector<IBlocksStreamPtr> children;
```

Comment:
It would hardly make a measurable difference, but it would still be nice to use a container that supports `pop_front()` for `O(1)`.